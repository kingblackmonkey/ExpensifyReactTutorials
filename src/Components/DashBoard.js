import React from 'react';
import WrapperListExpense  from './ListExpense'
import WrapperExpenseFilters from './ExpenseFIlters'


const DashBoard = ()=>{
    

    return(
   
        <div>
             <p>This Is Your Expenses</p>
             <WrapperExpenseFilters />
             <WrapperListExpense  />



        </div>
      
    )
  }
  
  export default DashBoard;