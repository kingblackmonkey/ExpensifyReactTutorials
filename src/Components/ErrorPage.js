import React from 'react';
import {Link} from 'react-router-dom';
const ErrorPage = ()=>{
    return (
        <div>
            <p>404!! not found</p>
            <Link to="/">Go Home</Link>
        </div>
        
  
    )
  }

  export default ErrorPage ; 