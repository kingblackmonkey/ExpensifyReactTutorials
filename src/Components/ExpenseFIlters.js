import React from 'react'
import {connect} from 'react-redux'
import {
    filterTextGenerator,  filterAmountGenerator, filterDateGenerator,
    filterStartDateGenerator, filterEndDateGenerator
} from '.././actions/filters'
import {DateRangePicker} from'react-dates'





// unconnected component now can have ability to dispatch action to change the 
// state of the store based on action
// return object from mapStateToProps get dispatch function added
// so your component can use dispatc function and change 
// the state
// then 
export class  ExpenseFilters extends React.Component {
   state = {
    focusedInput: null,
    startDate: this.props.filters.startDate,
    endDate : this.props.filters.endDate
   }


   onDatesChange = ({startDate, endDate })=>{


    
          this.setState(()=>(
            {
                startDate,
                endDate
                }
            ))
      // dispatch startDate filter
      this.props.filterStartDateGenerator({startDate});

    //   dispatch end date filter
      this.props.filterEndDateGenerator({endDate});        
  
   }

   onFocusChange=(focusedInput)=>{
   
    this.setState(()=>
        (
            {  
                focusedInput
            }
        )
        )

        
   }

   onChangeTextHandler = (e)=>{
    this.props.filterTextGenerator({text: e.target.value});
}

onChangeSelectHandler = (e)=>{
    e.target.value === "Amount"?
    this.props.filterAmountGenerator():
    this.props.filterDateGenerator();        
}

    render(){
            return (
                <div>
                        <input 
                        type="text" 
                        value = {this.props.filters.text}
                        onChange={this.onChangeTextHandler}/>

                    <select value = {this.props.filters.sortBy} onChange = {this.onChangeSelectHandler}>
                        <option value = "Date">Date</option>
                        <option value = "Amount">Amount</option>
                    </select>

                    <DateRangePicker
                    startDate={this.state.startDate} // momentPropTypes.momentObj or null,
                    startDateId="your_unique_start_date_id" // PropTypes.string.isRequired,
                    endDate={this.state.endDate} // momentPropTypes.momentObj or null,
                    endDateId="your_unique_end_date_id" // PropTypes.string.isRequired,
                    onDatesChange={this.onDatesChange} // PropTypes.func.isRequired,
                    focusedInput={this.state.focusedInput} // PropTypes.oneOf([START_DATE, END_DATE]) or null,
                    onFocusChange={this.onFocusChange} // PropTypes.func.isRequired,
                    numberOfMonths={1}
                    isOutsideRange={()=>{return false}}
                    showClearDates = {true}
                   />                    


                </div>
            ) 

        }
}
    

const mapStateToProps = (state)=>{
   
    return  {
        filters: state.filters
    }
}
 
   
const mapDispatchToProps = {
    filterStartDateGenerator,
    filterEndDateGenerator,
    filterStartDateGenerator,
    filterTextGenerator,
    filterAmountGenerator,
    filterDateGenerator
}




export default connect(mapStateToProps, mapDispatchToProps)(ExpenseFilters);