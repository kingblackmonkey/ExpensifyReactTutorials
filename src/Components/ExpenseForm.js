

import React from 'react';
import moment from 'moment'
import 'react-dates/lib/css/_datepicker.css';
import { SingleDatePicker } from 'react-dates';

window.moment = moment

class ExpenseForm  extends React.Component{
  // using ternary operator to set values for state
  // Expense component is being used in two component Add expense comp and
  // edit expense component
  // if it is in AddExpense component then no edit expense object
  // then default value will be string so user can enter values
  // if it is in EditExpense comp then there is edit expese object being  
  // passed down then we set values according to these values in edit 
  // object expense

  // the benefit of using state in this from component is that
  // you can validate user input in real time and transform input value in 
  // state with correct format  before you diispatch to redux

  // we dont dispatch add expense or remove expense in this form component
  // beacause we will resuse this component in edit component and in addexpense 
  // component. Each component will dispatch different action
  // if we dispatch add expense in this form compnent then we can not reuse this form
  // component to dispatch edit expense option 
    state = {
      description : this.props.expense? this.props.expense.description: '',
      amount: this.props.expense? (this.props.expense.amount / 100).toString(): '' ,
      note : this.props.expense? this.props.expense.note :'',
      timePayAt :this.props.expense?  moment( this.props.expense.timePayAt):moment(),
      focused : false,
      error: false
    }


    onchangeDescription = (e)=>{
      // console.log(e.target.value)
      const value = e.target.value ;
      this.setState ((state)=>{
        return {
          description: value
        }
      })
     
    }
  
    onAmountChange = (e)=>{
      const amount = e.target.value;
    
  
  //  set !amount so we can set state back to empty string so we can delete input number
  // when we delete number input feild it will be empty string and we make it true
  // by flipping it and set state on amount value to empty string so we can update number input
  // field with empty string as after we delete all value on the  input field
      if(!amount || /^\d{1,}(\.\d{0,2})?$/.test(amount)){
         
        this.setState((state)=>{
            return {
              amount
            }
        })
      }
    }
  
    onNoteChange = (e)=>{
      const value = e.target.value;
      this.setState ((state)=>{
        return {
          note: value
        }
      })
    }
  
    onDateChangeHandler = (timePayAt)=>{
     
     if(timePayAt){
         this.setState(()=>(
        {
          timePayAt
        }
      ))
     }
    
    }
  
    onFocusChangeHandler =({ focused })=>{
          this.setState(()=>(
            {
              focused
            }
          ))
    }
  
    onSubmitHandler = (e)=>{
        e.preventDefault();
      if(!this.state.description || !this.state.amount) {
        this.setState(()=>{
            return {
              error: true
            }
        });
      }else{

        this.props.onSubmitPassedUp(
          {
             description: this.state.description,
             amount: parseFloat(this.state.amount) * 100,
             note: this.state.note,
             timePayAt: this.state.timePayAt.valueOf()
          }
      ) 
      
      }
  
    }
    render(){
   
      return(  
      <div>
       
          {this.state.error && <p>Please Enter Description and Amount</p>}
          <form onSubmit = {this.onSubmitHandler}>
              <input 
              type='text' 
              value = {this.state.description} 
              autoFocus 
              placeholder='Description'
              onChange= {this.onchangeDescription}
              />
              
              <input 
              type='text' 
              placeholder = 'Amount'
              value = {this.state.amount}
              onChange = {this.onAmountChange}/> 
              
              
              
              <textarea 
              placeholder=' Note ' 
              value = {this.state.note}
              onChange = {this.onNoteChange}/> 
            <SingleDatePicker
                  date={this.state.timePayAt} // momentPropTypes.momentObj or null
                  onDateChange={this.onDateChangeHandler} // PropTypes.func.isRequired
                  focused={this.state.focused} // PropTypes.bool
                  onFocusChange={this.onFocusChangeHandler} // PropTypes.func.isRequired
                  numberOfMonths = {1}
                  isOutsideRange = {(day)=>{return false}}
                  id="your_unique_id" // PropTypes.string.isRequired,
                  />
  
  
  
              <button> Add Expense</button>
          </form>
  
  
      </div>
      )
    }
  
    
  
        
      
    }
  
    export default ExpenseForm;  
  