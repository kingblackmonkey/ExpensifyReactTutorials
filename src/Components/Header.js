import React from 'react';
import {NavLink} from 'react-router-dom';
const Header= ()=>{
  
  
    return (
   
    <header>
        <h1>Expense App</h1>
        <div>
            <NavLink  to="/"  exact = {true}  activeClassName="selected" >               
               DashBorad           
            </NavLink>
        </div>
       
        <div>
            <NavLink to={`/addExpense`} activeClassName="selected">           
               
         Add Expense
                 
            </NavLink>
        </div>
     
       
            
    </header>
  
    )
  }

  export default Header;