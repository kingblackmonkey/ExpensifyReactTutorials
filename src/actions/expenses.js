import uuidv4 from "uuid/v4"

// add-expense generator
export const addExpenseGenerator = (   
    {   
         description = '',   
          note = '' ,     
           amount = 0,     
            timePayAt = 0    
    } = {} )=>(   
    
        {
        type:"ADD_EXPENSE",
        expense: {
            id: uuidv4(),
            description,
            note,
            amount,
            timePayAt 
        }
       
    }
);

// remove-expense generator 
export const removeExpenseGenerator = ({id} = {})=>(
    {
        type: "REMOVE_EXPENSE",
        id
    }
);
   
// edit-expense generator
export const editExpenseGenerator = ({id, description, amount, note})=>(
    {
        type: "EDIT_EXPENSE",
        expense:{
            id,
            description,
            amount,
            note
        }
    }
);