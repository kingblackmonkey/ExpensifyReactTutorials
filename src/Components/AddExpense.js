import React from 'react';
import ExpenseForm from './ExpenseForm'
import {connect} from 'react-redux'
import { addExpenseGenerator} from '.././actions/expenses'



export const AddExpense = (props)=>(
  <div>
      <h1> Add Expense Page</h1>

      <ExpenseForm 
      onSubmitPassedUp = {(expense)=>{
        //  props.dispatch(addExpenseGenerator(expense))
      props.addExpenseGenerator(expense)

         props.history.push('/')
      }}
      />

  </div>
)

// const mapDispatchToProps = (dispatch)=>{
//   return {
//     onSubmit : (expense) => {return dispatch(addExpenseGenerator(expense))}
//   }
// }


const mapDispatchToProps = {addExpenseGenerator}

export default connect(null, mapDispatchToProps)(AddExpense);
