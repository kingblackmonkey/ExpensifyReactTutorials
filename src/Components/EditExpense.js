import React from 'react'
import ExpenseForm from './ExpenseForm'
import {connect} from 'react-redux'
import {editExpenseGenerator,removeExpenseGenerator} from '../actions/expenses'




// this named export here is for testing 
export const EditExpense= (props)=> {

    return (
    <div>
        <p>This is edit page</p>
        
        <ExpenseForm 
        
            expense = {props.expenses.find((expense)=>{
                return props.match.params.id === expense.id
           })} 
           
           
           onSubmitPassedUp={(expense)=>{
        //    dispatch 
            props.editExpenseGenerator({ ...props.expense ,...expense})

        // redirect
           props.history.push('/');
        //    console.log(expense)
        }}/>

        <button  onClick = {()=>{
            // dispatch to remove expense
            props.removeExpenseGenerator({id : props.match.params.id });
            // redirect
            props.history.push('/');

        }}>
            remove
        </button>
        
    </div>
    )
} 
    




const mapStateToProps = (state)=>{
    
       
          return {
           expenses:    state.expenses
        }
}

const mapDispatchToProps = {editExpenseGenerator, removeExpenseGenerator }
   
export default connect(mapStateToProps, mapDispatchToProps)(EditExpense) ; 
