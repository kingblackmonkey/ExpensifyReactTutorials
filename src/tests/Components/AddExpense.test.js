import React from 'react';
import {AddExpense } from '../../Components/AddExpense'
import { shallow} from 'enzyme';
import expenses from '../dummyExpenses/dummyExpenses'


// jest give life cycle method to  run each time before each test
// we can set up one test case once  and reuse it if you so we can just write assertion 
// it works if you finds you repeatedly set up the test case
// there is other life cycle methods too
// it is global 
// this global function will run once before each case

let  mockAddExpenseGenerator , mockHistory, wrapper 

beforeEach(()=>{
     mockAddExpenseGenerator = jest.fn();   
     mockHistory = {push:  jest.fn()}    
 
    wrapper = shallow(<AddExpense 
        addExpenseGenerator = {mockAddExpenseGenerator} 
        history = {mockHistory}

        />)
})
test ('should render add expense component ',()=>{
 
       
    //   wrapper.find('ExpenseForm').prop('onSubmitPassedUp')(expenses[0])

      expect (wrapper).toMatchSnapshot()   
      
})



test ('should render add expense component and dispatch and add expense',()=>{

       
     wrapper.find('ExpenseForm').prop('onSubmitPassedUp')(expenses[0])

      expect ( mockAddExpenseGenerator).toHaveBeenCalledWith(expenses[0])   
      expect ( mockHistory.push).toHaveBeenCalledWith('/')  
})