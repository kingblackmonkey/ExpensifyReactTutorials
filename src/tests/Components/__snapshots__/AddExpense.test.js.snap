// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render add expense component  1`] = `
<div>
  <h1>
     Add Expense Page
  </h1>
  <ExpenseForm
    onSubmitPassedUp={[Function]}
  />
</div>
`;
